src/parameterizations/vertical/MOM_diabatic_driver.F90

```diff
@@ -1575,6 +1616,108 @@ subroutine diabatic_ALE(u, v, h, tv, BLD, fluxes, visc, ADp, CDp, dt, Time_end,
     if (CS%id_Sdif > 0) call post_data(CS%id_Sdif, Sdif_flx, CS%diag)
   endif
 
+  if (CS%Use_Kd_Work_or_N2_diag) then
+
+     ! We could be more selective about allocating both of these, but the logic gets messy
+    allocate(bf%N2_salt(isd:ied,jsd:jed,nz+1), source=0.0)
+    allocate(bf%N2_temp(isd:ied,jsd:jed,nz+1), source=0.0)
+
+    ! Only allocate these as necessary
+    if (CS%do_bflx_salt) &
+      allocate(bf%Bflx_salt(isd:ied,jsd:jed,nz+1), source=0.0)
+    if (CS%do_bflx_temp) &
+      allocate(bf%Bflx_temp(isd:ied,jsd:jed,nz+1), source=0.0)
+    if (CS%id_Bdif_dz>0 .or. CS%id_Bdif_salt_dz>0 .or. CS%id_Bdif_dz_ePBL>0 .or. CS%id_Bdif_dz_ddiff_salt>0) &
+      allocate(bf%Bflx_salt_dz(isd:ied,jsd:jed,nz), source=0.0)
+    if (CS%id_Bdif_dz>0 .or. CS%id_Bdif_temp_dz>0 .or. CS%id_Bdif_dz_ePBL>0 .or. CS%id_Bdif_dz_ddiff_temp>0) &
+      allocate(bf%Bflx_temp_dz(isd:ied,jsd:jed,nz), source=0.0)
+
+    !Compute N2 and don't mask negatives here
+    EOSdom(:) = EOS_domain(G%HI)
+    !$OMP parallel do default(shared)
+    do j=js,je
+      if (associated(tv%p_surf)) then
+        do i=is,ie ; p_i(i) = tv%p_surf(i,j) ; enddo
+      else
+        do i=is,ie ; p_i(i) = 0.0 ; enddo
+      endif
+      do K=2,nz
+        do i=is,ie
+          p_i(i) = p_i(i) + (GV%g_Earth * GV%H_to_RZ) * h(i,j,k-1)
+        enddo
+        T_i = 0.5*(tv%T(:,j,k-1)+tv%T(:,j,k))
+        S_i = 0.5*(tv%S(:,j,k-1)+tv%S(:,j,k))
+        call calculate_density_derivs(T_i, S_i, p_i, dRhodT, dRhodS, tv%eqn_of_state, EOSdom)
+        do i=is,ie
+          I_dzval = 1.0 / (dz_neglect + 0.5*(dz(i,j,k-1) + dz(i,j,k)))
+          bf%N2_salt(i,j,K) = -(tv%S(i,j,k-1) - tv%S(i,j,k))*g_rho0*dRhodS(i) * I_dzval
```

Comment:
This expression is supposed to have units of [T-2 ~> s-2] to be consistent with its documented units and the conversion factor in the `register_diag_field()` call for "N2_salt_diabatic" at line 3397, but to me it looks like this has units of [L2 H-1 Z-1 T-2 ~> s-2 or m3 kg-1 s-2].  Please revise this expression and verify that this diagnostic is passing dimensional consistency testing.  Also, when in non-Boussinesq mode please try to avoid any dependencies on the Boussinesq reference density, as is currently embedded in `g_rho0`.

Reply:
Note to self, see wave speed for example

Reply:
This has been updated as suggested.